core/new-gui/src/app/workspace/service/workflow-graph/model/sync-texera-model.spec.ts

```diff
@@ -0,0 +1,667 @@
+
+import * as joint from 'jointjs';
+
+
+describe('SyncTexeraModel', () => {
+
+  let texeraGraph: WorkflowGraph;
+  let jointGraphWrapper: JointGraphWrapper;
+
+  function getAddOperatorValue(operator: OperatorPredicate) {
+    return {
+      operator: operator,
+      point: mockPoint
+    };
+  }
+
+  /**
+   * Returns a mock JointJS operator Element object (joint.dia.Element)
+   * The implementation code only uses the id attribute of the object.
+   *
+   * @param operatorID
+   */
+  function getJointOperatorValue(operatorID: string) {
+    return {
+      id: operatorID
+    };
+  }
+
+  /**
+   * Returns a mock JointJS Link object (joint.dia.Link)
+   * It includes the attributes and functions same as JointJS
+   *  and are used by the implementation code.
+   * @param link
+   */
+  function getJointLinkValue(link: OperatorLink) {
+    // getSourceElement, getTargetElement, and get all returns a function
+    //  that returns the corresponding value
+    return {
+      id: link.linkID,
+      attributes: {
+        source: { id: link.source.operatorID, port: link.source.portID },
+        target: { id: link.target.operatorID, port: link.target.portID }
+      }
+      // getSourceElement: () => ({ id: link.source.operatorID }),
+      // getTargetElement: () => ({ id: link.target.operatorID }),
+      // get: (port: string) => {
+      //   if (port === 'source') {
+      //     return { port: link.source.portID };
+      //   } else if (port === 'target') {
+      //     return { port: link.target.portID };
+      //   } else {
+      //     throw new Error('getJointLinkValue: mock is inconsistent with implementation');
+      //   }
+      // }
+    };
+  }
+
+  /**
+   * This helper function returns a mock JointJS link object (joint.dia.Link)
+   *  that is only connected to a source port, but detached from the target port.
+   *
+   * This scenario happens when the user is still moving the link
+   *  and it is not connected to a target port.
+   *
+   * @param link an operator link, but the target operator and target link is ignored
+   */
+  function getIncompleteJointLink(link: OperatorLink) {
+    // getSourceElement, getTargetElement, and get all returns a function
+    //  that returns the corresponding value
+    return {
+      id: link.linkID,
+      getSourceElement: () => ({ id: link.source.operatorID }),
+      getTargetElement: () => null,
+      get: (port: string) => {
+        if (port === 'source') {
+          return { port: link.source.portID };
+        } else if (port === 'target') {
+          return null;
+        } else {
+          throw new Error('getJointLinkValue: mock is inconsistent with implementation');
+        }
+      }
+    };
+  }
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      providers: [
+      ]
+    });
+
+    texeraGraph = new WorkflowGraph();
+    jointGraphWrapper = new JointGraphWrapper(new joint.dia.Graph());
+  });
+
+  /**
```

Comment:
shouldn't there be a "only one link should be added when link addition occurs" and "only one link and no other link should be deleted when link deletion occurs"

Reply:
IMO they are sort of trival and there are many other test cases about links to make sure that is correct